typescript/schedules.md

```diff
@@ -0,0 +1,1246 @@
+    // delete schedule.spec.intervals;
+
+    // return false to stop retrying early
+  },
+  { retry: retryPolicy }
+);
+
+await schedule.trigger();
+await schedule.backfill({ startAt: new Date(), endAt: addWeeks(new Date(), 1) }); // also takes array
+await schedule.pause('note: pausing');
+await schedule.unpause('now unpause');
+
+await schedule.delete();
+
+const { schedules, nextPageToken } = await client.listByPage({
+  pageSize: 50,
+  nextPageToken: 'base64',
+});
+
+for await (const schedule: Schedule of client.list()) {
+  const { id, memo, searchAttributes } = schedule;
+  // ...
+}
+```
+
+### Types
+
+```ts
+import { DataConverter, LoadedDataConverter } from '@temporalio/common';
+import { loadDataConverter } from '@temporalio/internal-non-workflow-common';
+import {
+  composeInterceptors,
+  Headers,
+  Replace,
+  RetryPolicy,
+  SearchAttributes,
+  Workflow,
+} from '@temporalio/internal-workflow-common';
+import { temporal } from '@temporalio/proto';
+import os from 'os';
+import { Connection } from './connection';
+import { ScheduleClientCallsInterceptor, ScheduleClientInterceptors } from './interceptors';
+import { ConnectionLike, Metadata, WorkflowService } from './types';
+import { WorkflowHandle, WorkflowStartOptions } from './workflow-client';
+
+import type { NonNegativeInteger, RequireAtLeastOne } from 'type-fest';
+
+// TODO is a non-generic NonNegativeInteger possible? The ones below error due to no type argument
+export type PositiveInteger = Exclude<NonNegativeInteger, 0>;
+
+/**
+ * @format number of milliseconds or {@link https://www.npmjs.com/package/ms | ms-formatted string}
+ */
+export type Milliseconds = string | NonNegativeInteger;
+
+/**
+ * @format number of seconds or {@link https://www.npmjs.com/package/ms | ms-formatted string}
+ */
+export type Seconds = string | NonNegativeInteger;
+
+export interface UpdateScheduleOptions {
+  /**
+   * @default TODO
+   */
+  retry?: RetryPolicy;
+}
+
+export interface Backfill {
+  /** Time range to evaluate Schedule in. */
+  startAt: Date;
+  endAt: Date;
+
+  /** Override Overlap Policy for this request. */
+  overlapPolicyOverride?: ScheduleOverlapPolicy;
+}
+
+/**
+ * Handle to a single Schedule
+ */
+export interface ScheduleHandle {
+  /**
+   * This Schedule's identifier
+   */
+  readonly id: string;
+
+  /**
+   * Update the Schedule
+   */
+  update(mutateFn: (schedule: ScheduleDescription) => void | false, options?: UpdateScheduleOptions): Promise<void>;
+
+  /**
+   * Delete the Schedule
+   */
+  delete(): Promise<void>;
+
+  /**
+   * Trigger an Action to be taken immediately
+   */
+  trigger(): Promise<void>;
+
+  /**
+   * Run though the specified time period(s) and take Actions as if that time passed by right now, all at once. The
+   * Overlap Policy can be overridden for the scope of the Backfill.
+   */
+  backfill(options: Backfill | Backfill[]): Promise<void>;
+
+  /**
+   * Pause the Schedule
+   */
+  pause(note?: string): Promise<void>;
```

Comment:
note should be required (and must be non-empty)

Reply:
I'm thinking leaving optional and adding default note:

`@default "paused"`

Reply:
@dnr why is this important?
Can SDK supply a default?

Reply:
Basically, because "paused" is a bit of dynamic state in your production system (i.e. it can't be synced from a git repo). It should always be possible to attach some context or comments to state and configuration. For stuff that's more static, you can comment the values in code. for dynamic stuff like this, there should be a way to link to a slack thread or incident page or a person's name or something. Otherwise someone is going to look at the schedules page and see something paused and have to wonder "why is this paused? should it be? what's the criteria for when I can start it again?"  Of course we can't force users to have good operational practices, so maybe a default is fine, but we can at least encourage it.

Reply:
I'd be fine with a default for the lazy. No strong opinion.
The default should probably be "paused by TypeScript SDK" or something like that.